src/main/java/org/folio/list/services/refresh/ListRefreshService.java

```diff
@@ -61,22 +64,57 @@ public void doAsyncRefresh(ListEntity list, ShutdownTask shutdownTask) {
   @Async
   @Transactional(propagation = Propagation.NOT_SUPPORTED)
   public void doAsyncSorting(ListEntity list, UUID queryId, ShutdownTask shutdownTask) {
+    try (var autoCloseMe = shutdownTask) {
+      log.info("Performing async sorting for list {}, refreshId {}", list.getId(),
+        list.getInProgressRefreshId().map(UUID::toString).orElse("NONE"));
+      waitForQueryCompletion(list, queryId, shutdownTask);
+    } catch (Exception exception) {
+      log.error("Unexpected error when performing async refresh for list with id " + list.getId()
+        + ", refreshId " + (list.getInProgressRefreshId().map(UUID::toString).orElse("NONE")), exception);
+      refreshFailedCallback.accept(list, exception);
+    }
+  }
+
+  // @Async?
+  @Transactional(propagation = Propagation.NOT_SUPPORTED)
+  private void waitForQueryCompletion(ListEntity list, UUID queryId, ShutdownTask shutdownTask) throws InterruptedException {
+    log.info("Waiting for completion of query {} for list {}", queryId, list.getId());
+    QueryDetails queryDetails = queryClient.getQuery(queryId);
+    while (queryDetails.getStatus() == QueryDetails.StatusEnum.IN_PROGRESS) {
+      TimeUnit.SECONDS.sleep(GET_QUERY_TIME_DELAY);
```

Comment:
This will almost certainly get shot down by Sonarqube 😦 The approach Sonar recommends is a library called Awaitility and using it to do something like `await().atMost(2, Duration.SECONDS).until(() -> queryDetails.getStatus() != QueryDetails.StatusEnum.IN_PROGRESS));`

Afaik, we don't already have a dependency on Awaitility (except maybe as a shaded dependency in testcontainers, but we don't want to depend on a shaded dependency), so we'll need to add it.